contracts/contracts/strategies/NativeStaking/ValidatorRegistrator.sol

```diff
@@ -0,0 +1,239 @@
+        REGISTERED, // validator is registered on the SSV network
+        STAKED, // validator has funds staked
+        EXITING, // exit message has been posted and validator is in the process of exiting
+        EXIT_COMPLETE // validator has funds withdrawn to the EigenPod and is removed from the SSV
+    }
+
+    event RegistratorChanged(address newAddress);
+    event ETHStaked(bytes pubkey, uint256 amount, bytes withdrawal_credentials);
+    event SSVValidatorRegistered(bytes pubkey, uint64[] operatorIds);
+    event SSVValidatorExitInitiated(bytes pubkey, uint64[] operatorIds);
+    event SSVValidatorExitCompleted(bytes pubkey, uint64[] operatorIds);
+
+    /// @dev Throws if called by any account other than the Registrator
+    modifier onlyRegistrator() {
+        require(
+            msg.sender == validatorRegistrator,
+            "Caller is not the Registrator"
+        );
+        _;
+    }
+
+    /// @dev Throws if called by any account other than the Strategist
+    modifier onlyStrategist() {
+        require(
+            msg.sender == IVault(VAULT_ADDRESS).strategistAddr(),
+            "Caller is not the Strategist"
+        );
+        _;
+    }
+
+    /// @param _wethAddress Address of the Erc20 WETH Token contract
+    /// @param _vaultAddress Address of the Vault
+    /// @param _beaconChainDepositContract Address of the beacon chain deposit contract
+    /// @param _ssvNetwork Address of the SSV Network contract
+    constructor(
+        address _wethAddress,
+        address _vaultAddress,
+        address _beaconChainDepositContract,
+        address _ssvNetwork
+    ) {
+        WETH_TOKEN_ADDRESS = _wethAddress;
+        BEACON_CHAIN_DEPOSIT_CONTRACT = _beaconChainDepositContract;
+        SSV_NETWORK_ADDRESS = _ssvNetwork;
+        VAULT_ADDRESS = _vaultAddress;
+    }
+
+    /// @notice Set the address of the registrator which can register, exit and remove validators
+    function setRegistrator(address _address) external onlyGovernor {
+        emit RegistratorChanged(_address);
+        validatorRegistrator = _address;
+    }
+
+    /// @notice Stakes WETH to the node validators
+    /// @param validators A list of validator data needed to stake.
+    /// The `ValidatorStakeData` struct contains the pubkey, signature and depositDataRoot.
+    /// Only the registrator can call this function.
+    function stakeEth(ValidatorStakeData[] calldata validators)
+        external
+        onlyRegistrator
+        whenNotPaused
+    {
+        uint256 requiredETH = validators.length * 32 ether;
+
+        // Check there is enough WETH from the deposits sitting in this strategy contract
+        require(
+            requiredETH <= IWETH9(WETH_TOKEN_ADDRESS).balanceOf(address(this)),
+            "insufficient WETH"
+        );
+
+        // Convert required ETH from WETH
+        IWETH9(WETH_TOKEN_ADDRESS).withdraw(requiredETH);
+
+        // For each validator
+        for (uint256 i = 0; i < validators.length; ) {
```

Comment:
`validators.length` should be cached outside of the loop.

Reply:
thanks, fixed here: https://github.com/OriginProtocol/origin-dollar/pull/2053/commits/fd4f2d24d5db8723e7eef9ed8630a4f221684749